test/functional/plugin/lsp/completion_spec.lua

```diff
@@ -278,4 +288,218 @@ describe('vim.lsp._completion', function()
+      completion_result
+    )
+  end
+
+  local function get_matches()
+    return exec_lua('return _G.capture.matches')
+  end
+
+  --- @param pos { [1]: integer, [2]: integer }
+  local function trigger_completion_at_pos(pos)
+    exec_lua(
+      [[
+      local win = vim.api.nvim_get_current_win()
+      vim.api.nvim_win_set_cursor(win, ...)
+      vim.lsp.completion.trigger_completion()
+    ]],
+      pos
+    )
+
+    retry(nil, nil, function()
+      return exec_lua('return _G.capture.col') ~= nil
+    end)
+  end
+
+  it('fetches completions and shows them using complete on trigger_completion', function()
+    create_server({
+      isIncomplete = false,
+      items = {
+        {
+          label = 'hello',
+        },
+      },
+    })
+
+    feed('ih')
+    trigger_completion_at_pos({ 1, 1 })
+
+    eq({
+      {
+        abbr = 'hello',
+        dup = 1,
+        empty = 1,
+        icase = 1,
+        info = '',
+        kind = 'Unknown',
+        menu = '',
+        user_data = {
+          nvim = {
+            lsp = {
+              client_id = 1,
+              completion_item = {
+                label = 'hello',
+              },
+            },
+          },
+        },
+        word = 'hello',
+      },
+    }, get_matches())
+  end)
+
+  it('merges results from multiple clients', function()
+    create_server({
+      isIncomplete = false,
+      items = {
+        {
+          label = 'hello',
+        },
+      },
+    })
+    create_server({
+      isIncomplete = false,
+      items = {
+        {
+          label = 'hallo',
+        },
+      },
+    })
+
+    feed('ih')
+    trigger_completion_at_pos({ 1, 1 })
+
+    local matches = get_matches()
+    eq(2, #matches)
+    eq('hello', matches[1].word)
+    eq('hallo', matches[2].word)
+  end)
+
+  it('executes commands', function()
+    local completion_list = {
+      isIncomplete = false,
+      items = {
+        {
+          label = 'hello',
+          command = {
+            arguments = { '1', '0' },
+            command = 'dummy',
+            title = '',
+          },
+        },
+      },
+    }
+    local client_id = create_server(completion_list)
+
+    exec_lua(
+      [[
+      _G.called = false
+      local client = vim.lsp.get_client_by_id(...)
+      client.commands.dummy = function ()
+        _G.called = true
+      end
+    ]],
+      client_id
+    )
+
+    feed('ih')
+    trigger_completion_at_pos({ 1, 1 })
+
+    exec_lua(
+      [[
+      local client_id, item = ...
+      vim.v.completed_item = {
+        user_data = {
+          nvim = {
+            lsp = {
+              client_id = client_id,
+              completion_item = item
+            }
+          }
+        }
+      }
+
+      vim.api.nvim_exec_autocmds('CompleteDone', { buffer = bufnr })
+    ]],
+      client_id,
+      completion_list[1]
+    )
+
+    local matches = get_matches()
+    eq(1, #matches)
+    eq('hello', matches[1].word)
+    eq(true, exec_lua('return _G.called'))
```

Comment:
The new tests are all passing except for this which fails here. @mfussenegger I would appreciate your wisdom here 🥲

Reply:
Nvm, fixed!